src/main/kotlin/com/team4099/robot2020/subsystems/Feeder.kt

```diff
@@ -0,0 +1,152 @@
+
+    // take this out after adding one ballCount in superstructure to work with intake
+    var ballCount = 0
+
+    var feederState = FeederState.IDLE
+        set(value) {
+            if (value != field) {
+                HelixEvents.addEvent("FEEDER", "state changed to $value")
+            }
+        }
+    private var feederInPower = 0.0
+        set(value) {
+            if (value != field) {
+                inMasterSparkMax.set(value)
+            }
+            field = value
+        }
+    private var feederStopperPower = 0.0
+        set(value) {
+            if (value != field) {
+                stopperTalon.set(ControlMode.PercentOutput, value)
+            }
+            field = value
+        }
+
+    init {
+        inMasterSparkMax.inverted = false
+        inSlaveSparkMax.inverted = true
+
+        stopperTalon.inverted = true
+    }
+
+    enum class FeederState {
+        HOLD, INTAKE, SHOOT, EXHAUST, IDLE
+    }
+
+    override fun outputTelemetry() {
+        SmartDashboard.putString("feeder/feederState", feederState.toString())
+        SmartDashboard.putNumber("feeder/feederInPower", feederInPower)
+        SmartDashboard.putNumber("feeder/feederStopperPower", feederStopperPower)
+    }
+
+    override fun checkSystem() {}
+
+    override fun registerLogging() {
+        HelixLogger.addSource("Feeder Master Motor Power") { inMasterSparkMax.appliedOutput }
+        HelixLogger.addSource("Feeder Slave Motor Power") { inSlaveSparkMax.appliedOutput }
+        HelixLogger.addSource("Feeder Stopper Motor Power") { stopperTalon.motorOutputPercent }
+
+        HelixLogger.addSource("Feeder Master Motor Current") { inMasterSparkMax.outputCurrent }
+        HelixLogger.addSource("Feeder Slave Motor Current") { inSlaveSparkMax.outputCurrent }
+        HelixLogger.addSource("Feeder Stopper Motor current") { stopperTalon.supplyCurrent }
+
+        HelixLogger.addSource("Feeder State") { feederState }
+    }
+
+    override fun zeroSensors() { }
+
+    override fun onStart(timestamp: Double) {
+        feederState = FeederState.HOLD
+    }
+
+    @Synchronized
+    override fun onLoop(timestamp: Double, dT: Double) {
+
+        if (inBeamBroken) {
+            if (inBeamBrokenTimestamp == -1.0) {
+                inBeamBrokenTimestamp = timestamp
+            }
+        } else {
+            if (timestamp - inBeamBrokenTimestamp >= Constants.BeamBreak.IN_BEAM_BROKEN_BALL_TIME) {
+                ballCount++
+            }
+            inBeamBrokenTimestamp = -1.0
+        }
+
+        if (outBeamBroken) {
+            if (outBeamBrokenTimestamp == -1.0) {
+                outBeamBrokenTimestamp = timestamp
+            }
+        } else {
+            if (timestamp - outBeamBrokenTimestamp >= Constants.BeamBreak.OUT_BEAM_BROKEN_BALL_TIME) {
+                ballCount--
+            }
+            outBeamBrokenTimestamp = -1.0
+        }
+
+        when (feederState) {
+            FeederState.INTAKE -> {
```

Comment:
Need something to make the feeder not just run continuously in the standard `INTAKE` state. If we do this, the first ball we intake will go all the way up the feeder and potentially get stuck. If we have a beam break at the front of the feeder, and it's far enough forward and reacts very quickly, we could do something like just running the intake when that beam is broken, and only run until the ball clears the front of the feeder or something. Also could do something w/ intake current sensing as a secondary check to run the feeder